File: packages/go/firmware_test.go
package sonicwall

import "testing"

func TestFirmwareLimitationReason(t *testing.T) {
	tests := []struct {
		status  int
		message string
		want    FirmwareLimitationReason
	}{
		{404, "API not found", FirmwareAPINotFound},
		{400, "API endpoint is incomplete", FirmwareEndpointIncomplete},
		{400, "incomplete", FirmwareEndpointIncomplete},
		{404, "command xyz not found", FirmwareCommandNotFound},
		{405, "Non config mode", FirmwareNonConfigMode},
	}
	for _, tt := range tests {
		got := firmwareLimitationReason(tt.status, tt.message)
		if got == nil || *got != tt.want {
			t.Errorf("firmwareLimitationReason(%d, %q) = %v, want %q", tt.status, tt.message, got, tt.want)
		}
	}
	if firmwareLimitationReason(500, "Internal server error") != nil {
		t.Error("expected nil for generic 500")
	}
}

func TestNewHTTPErrorUnsupportedEndpoint(t *testing.T) {
	body := &SonicOSErrorResponse{
		Status: SonicOSStatus{
			Success: false,
			Info: []SonicOSStatusInfo{{
				Code:    400,
				Message: "API endpoint is incomplete",
			}},
		},
	}
	err := newHTTPError(400, body)
	if !IsUnsupportedEndpoint(err) {
		t.Fatalf("expected UnsupportedEndpointError, got %T", err)
	}
	ue := err.(*UnsupportedEndpointError)
	if ue.Reason != FirmwareEndpointIncomplete {
		t.Fatalf("reason = %q, want %q", ue.Reason, FirmwareEndpointIncomplete)
	}
}

File: packages/typescript/tests/firmware.test.ts
import { describe, expect, it } from "vitest";
import { firmwareLimitationReason, isFirmwareUnsupportedError } from "../src/firmware.ts";
import {
  SonicWallHTTPError,
  UnsupportedEndpointError,
  raiseForSonicOSBody,
} from "../src/errors.ts";

describe("firmwareLimitationReason", () => {
  it.each([
    [404, "API not found", "api_not_found"],
    [400, "API endpoint is incomplete", "endpoint_incomplete"],
    [400, "incomplete", "endpoint_incomplete"],
    [404, "command xyz not found", "command_not_found"],
    [405, "Non config mode", "non_config_mode"],
    [500, "Internal server error", null],
  ] as const)("classifies %i %s as %s", (statusCode, message, expected) => {
    expect(firmwareLimitationReason(statusCode, message)).toBe(expected);
  });
});

describe("UnsupportedEndpointError", () => {
  it("is raised for endpoint incomplete responses", () => {
    expect(() =>
      raiseForSonicOSBody(400, {
        status: {
          success: false,
          info: [{ code: 400, message: "API endpoint is incomplete" }],
        },
      })
    ).toThrow(UnsupportedEndpointError);

    try {
      raiseForSonicOSBody(400, {
        status: {
          success: false,
          info: [{ code: 400, message: "API endpoint is incomplete" }],
        },
      });
    } catch (err) {
      expect(err).toBeInstanceOf(UnsupportedEndpointError);
      expect((err as UnsupportedEndpointError).reason).toBe("endpoint_incomplete");
    }
  });

  it("isFirmwareUnsupportedError detects typed errors", () => {
    const exc = new UnsupportedEndpointError(400, "API endpoint is incomplete", "endpoint_incomplete");
    expect(isFirmwareUnsupportedError(exc)).toBe(true);
    expect(isFirmwareUnsupportedError(new SonicWallHTTPError(500, "fail"))).toBe(false);
  });
});
